src/modules/iotjs_module_mqtt.c

```diff
@@ -0,0 +1,777 @@
+        return JS_CREATE_ERROR(COMMON, "MQTT: PUBREL packet is corrupted");
+      }
+
+      iotjs_mqtt_message_response_t *pubrel =
+          (iotjs_mqtt_message_response_t *)buffer;
+
+      char control_packet_reserved = first_byte & 0b00001111;
+      if (control_packet_reserved != 2) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: PUBREL packet is corrupted");
+      }
+
+      IOTJS_UNUSED(pubrel);
+      iotjs_create_ack_callback(buffer, IOTJS_MAGIC_STRING_ONPUBREL, jsref);
+      break;
+    }
+    case PUBCOMP: {
+      if (remaining_length != 2) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: PUBCOMP packet is corrupted");
+      }
+
+      iotjs_create_ack_callback(buffer, IOTJS_MAGIC_STRING_ONPUBCOMP, jsref);
+      break;
+    }
+    case SUBACK: {
+      // We assume that only one topic was in the SUBSCRIBE packet.
+      if (remaining_length != 3) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: SUBACK packet is corrupted");
+      }
+
+      iotjs_mqtt_message_suback_t *suback =
+          (iotjs_mqtt_message_suback_t *)buffer;
+
+      if (suback->return_code == 128) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: Subscription was unsuccessful");
+      }
+
+      // The callback takes the granted QoS as parameter.
+      iotjs_jargs_t args = iotjs_jargs_create(1);
+      iotjs_jargs_append_number(&args, suback->return_code);
+
+      jerry_value_t sub_fn =
+          iotjs_jval_get_property(jsref, IOTJS_MAGIC_STRING_ONSUBACK);
+      iotjs_make_callback(sub_fn, jsref, &args);
+      jerry_release_value(sub_fn);
+      iotjs_jargs_destroy(&args);
+      break;
+    }
+    case UNSUBACK: {
+      if (remaining_length != 2) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: UNSUBACK packet is corrupted");
+      }
+
+      iotjs_create_ack_callback(buffer, IOTJS_MAGIC_STRING_ONUNSUBACK, jsref);
+      break;
+    }
+    case PINGRESP: {
+      if (remaining_length != 0) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: PingRESP packet is corrupted");
+      }
+      jerry_value_t fn =
+          iotjs_jval_get_property(jsref, IOTJS_MAGIC_STRING_ONPINGRESP);
+      iotjs_make_callback(fn, jsref, iotjs_jargs_get_empty());
+      jerry_release_value(fn);
+      break;
+    }
+    case DISCONNECT:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Broker disconnected the client");
+
+    case CONNECT:
+    case SUBSCRIBE:
+    case UNSUBSCRIBE:
+    case PINGREQ:
+      return JS_CREATE_ERROR(
+          COMMON, "MQTT: Unallowed packet has arrived to the client");
+  }
+
+  return jerry_create_undefined();
+}
+
+
+JS_FUNCTION(MqttSubscribe) {
+  DJS_CHECK_THIS();
+  DJS_CHECK_ARGS(1, object);
+
+  jerry_value_t joptions = JS_GET_ARG(0, object);
+
+  jerry_value_t jtopic =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_TOPIC);
+  jerry_value_t jqos =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_QOS);
+
+  uint8_t qos = 0;
+  if (!jerry_value_is_undefined(jqos) && !jerry_value_has_error_flag(jqos)) {
```

Comment:
I think there is a utility now which converts strings and buffers to char*, which is exactly what you need here. It also returns with error if the conversion is not possible.